rfcs/2023-02-08-7496-vrl-return.md

```diff
@@ -0,0 +1,72 @@
+# RFC 7496 - 2023-02-02 - VRL Return Keyword
+
+Add return expression to the Vector Remap Language.
+
+## Context
+
+- #7496
+
+## Cross cutting concerns
+
+- None.
+
+## Scope
+
+### In scope
+
+- Adding a return expression to VRL.
+
+### Out of scope
+
+- Adding new keywords for similar purposes such as `drop`.
+- Defining semantics of keywords that are usually used for other purposes in other languages such as `break`.
+- Defining parameterized return, e.g. `return parse_json!(.message)`. The proposed return expression always passes current `.` to output and takes no input.
+
+## Pain
+
+- Aborting with changing the input cannot be easily done.
+- VRL code is often unnecessarily indented because of lack of early returns.
+
+## Proposal
+
+### User Experience
+
+- A `return` expression causes the VRL program to terminate, keeping any modifications made to the event.
+- The keyword is used by itself with no parameters and `.` will be sent to the output.
```

Comment:
Worth considering returning an `expression`, it can be `.` or a variable or something that resolves to a constant. What do you think?

Reply:
Yes, I think that is the end l goal to go for. I've tried to keep the scope small, but it can also be part of the RFC, with the current scope being probably the first implementation and having positional arguments being enhancement of that.

I opted for omitting it since it felt to me that writing
```
. = expr
return
```
is only slightly worse than
```
return expr
```

Do you think it's better to have it in this RFC or having it split as a later enhancement?

Reply:
Also I think even if expressions are allowed, they should be optional, allowing simple `return` instead of the more explicit `return .`

But now I see that could be a contentious decision and some people can prefer that the language wants them to be explicit, so I'd like to also know your opinion on this.

Reply:
I would take inspiration from https://doc.rust-lang.org/std/keyword.return.html so I would like to return an expression. See [here](https://github.com/vectordotdev/vector/blob/v0.36.0/src/conditions/vrl.rs#L101-L106), we can use both the returned value and the target `.`.

Reply:
To clarify a bit more: would `return foo` use the value of `foo` as the event to emit? Or would it still use `.`? If the latter, what would the "returned value" be used for?

Reply:
> I would take inspiration from https://doc.rust-lang.org/std/keyword.return.html so I would like to return an expression. See [here](https://github.com/vectordotdev/vector/blob/v0.36.0/src/conditions/vrl.rs#L101-L106), we can use both the returned value and the target `.`.

I'm not sure what you mean, would you like to have only `return expr` and no `return` without arguments?

I think there will always be a difference between Rust and VRL because the latter already implicitly uses `.` when returning at the end of a program while the former implicitly returns the result of the last expression at the end of a function. So in VRL, it makes much more sense to me to have a parameter-less return, returning the same thing as if it were the end of the program.

> To clarify a bit more: would `return foo` use the value of `foo` as the event to emit? Or would it still use `.`? If the latter, what would the "returned value" be used for?

I would expect it to emit `foo`.

Reply:
> > To clarify a bit more: would `return foo` use the value of `foo` as the event to emit? Or would it still use `.`? If the latter, what would the "returned value" be used for?
> 
> I would expect it to emit `foo`.

Likewise. I'd suggest we could allow `return` both with and without an "argument". If an "argument" is provided, then it would return that, otherwise it would return `.`. My thinking is that `return X` could be useful for VRL conditions that have to evaluate to a boolean (where you might want to `return true` or `return false` early) whereas just `return` would be more common in `remap` programs where you emit an entire event.